Added in input_asure.py to locidex merge

sgsutcliffe opened the pull request:
### STRY0017229

Add method to provide a sample name associated with an MLST JSON file in "locidex merge"

### Description

We want to be able to provide a sample name/identifier alongside each MLST JSON file in "locidex merge" so I can override the sample name stored within the JSON file without writing a new JSON file prior to "locidex merge".

The subtext of the added feature is it will allow pipelines like `gasnomenclature` to be more efficient, by integrating the check step at the same time the MLST file is already being read, and written together.

### Goals

- [x] The "locidex merge" command will be modified to override the sample name stored within the MLST JSON file.
                - One option is to use a syntax like "SAMPLE=FILE" when passing on the command line.
                - Another option is to pass a CSV file as input with two columns "sample,file"
- [x] When a sample name is provided to "locidex merge", this will override the sample name stored in the MLST JSON file.
- [x] The output "profiles.csv" file will use the overridden sample names as identifiers if they were provided, otherwise it will use sample identifiers in the MLST JSON file.
- [x] The "locidex merge" command will be modified to output an "error.csv" file with any warnings or errors in input MLST JSON files, one sample/file per row. This will match format of existing error file produced by "INPUT_ASSURE" in gasnomenclature pipeline.
- [x] The updated "locidex merge" command will be backwards compatible with the previous functionality (where the sample name is derived from the contents of the MLST JSON file).
- [x] Unit/integration tests should be developed to confirm new functionality and existing functionality (overriding sample names in MLST JSON files).

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+        sampledict = {}
+        for line in f:
+            line = line.rstrip().split(",")
+            if len(line) != 2:
```
Can you remove the magic number? e.g. Assign the numbers to variables above with a descriptive name. I assume the "2" is supposed to refer to the expected columns length

sgsutcliffe replied:
*poof* went away (thanks to a magical panda) 468f76e

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+                logger.critical("File should be tsv with two columns [sample,mlst_alleles]")
+                raise_file_not_found_e(logger)
+            sample = line[0]
+            mlst_file = os.path.basename(line[1])
```
Same as above, I understand the intention of the zero but what does the "1" refer to?

sgsutcliffe replied:
magical panda strikes again 468f76e

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+
+def compare_profiles(mlst, sample_id, file_name):
+    # Extract the profile from the json_data
+    profile = mlst.get("data", {}).get("profile", {})
```
If a string is repeated can you assign it to a variable?  

Just to make refactoring easier in the future and minimize errors due to spelling mistakes in the future.

sgsutcliffe replied:
23452cb

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+        error_message = (
+            f"{file_name} is missing the 'profile' section or is completely empty!"
+        )
+        print(error_message)
```
For the error message could you use the logger and raise the correct error? There is an error handler in the main loop that will handle your error for example you could add `raise KeyError(error_message). 

One reason for using the logger over print is that print is buffered to stdout which may not always be collected, where as the log messages write to stderr which is not buffered. Additionally we can format the log messages in a pretty way. simply replace the print statement with `logger.error(error_message)` 


If you are curious about the error handler you can see how the error handler is implemented here: https://github.com/phac-nml/locidex/blob/eedba57e561f96cfcabba009c3c7cca062fefc0b/locidex/main.py#L42-L54 and for how to call the logger

sgsutcliffe replied:
Good catch, and that makes sense to me! In line with some of the comments Aaron made which is adapted what was a single input_assure script that was used per process in the nextflow pipeline, to proper integration with the locidex. 4631d27

sgsutcliffe replied:
This comment was extra helpful because it reminded me that I forgot to add a new output which is the corrected MLST file gzipped!

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+            f"{file_name} is missing the 'profile' section or is completely empty!"
+        )
+        print(error_message)
+        sys.exit(1)
```
Remove the sys.exit call, the error handler will handle the outputs.

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+    elif len(keys) > 1:
+        # Check if sample_id matches any key
+        if not match_status:
+            error_message = f"No key in the MLST JSON file ({file_name}) matches the specified sample ID '{sample_id}'. The first key '{original_key}' has been forcefully changed to '{sample_id}' and all other keys have been removed."
```
For code below see the above comment about using strings. It is best to assign the strings to variables. e.g. "data" and "profile"

mattheww95 commented on locidex/merge.py:
```diff
@@ -198,18 +258,33 @@ def run_merge(config):
+
+    #Write error messages for profile mismatch (compare_profiles())
+    for error_message in compare_error:
+            if error_message[2]:
```
magic numbers 🧙

sgsutcliffe replied:
*poof* dissappeared with c9b0d94

mattheww95 commented on locidex/merge.py:
```diff
@@ -198,18 +258,33 @@ def run_merge(config):
+                with open(output_error_file, "w", newline="") as f:
+                    writer = csv.writer(f)
+                    writer.writerow(["sample", "JSON_key", "error_message"])
+                    writer.writerow([error_message[0], error_message[1], error_message[2]])
```
magic numbers 🧙

sgsutcliffe replied:
*poof* disappeared with c9b0d94

mattheww95 commented on locidex/merge.py:
```diff
@@ -198,18 +258,33 @@ def run_merge(config):
-    
+
+    #Write error messages for profile mismatch (compare_profiles())
+    for error_message in compare_error:
```
Should the program raise a non-zero exit status if errors are encountred so the pipeline notifies the user that something went wrong? 

it may be useful for pipelines as they will not die nessecarily as all outputs will be present, but Nextflow will notify the user of a non-zero exit status.

sgsutcliffe replied:
I think I might rename these, as "error" is not really what they are (i.e. non-zero exit would not be right here). We're collapsing `input_assure` (a process from gasnomenclature pipeline) into locidex. `input_assure` was a way to override the mlst profile locidex merge uses in the output file. If someone supplies a renaming file we have to assume they will be aware that the output would be modified, we simply want to provide a file that explains how the file was modified. We called this an error report but its more of a warning, or friendly notification.

sgsutcliffe replied:
This comment reminded me I need to make a new output file with the modified MLST json file!

sgsutcliffe replied:
I changed the variable names so it is a bit more clear [580c272](https://github.com/phac-nml/locidex/pull/39/commits/580c272884067dd5354fbb43ff15ce7410fbd565)

sgsutcliffe replied:
Changed (c9b0d94) the variable to `modified_MLST_file_list` which is less urgent or bad as we would expect from `error`:
```python
    #Write report of all the MLST files with profile mismatch and how MLST profiles with mismatch were modified
    df = pd.DataFrame(modified_MLST_file_list)
    df.to_csv(f'{outdir}/MLST_error_report.csv', index=False, header=False)
```

mattheww95 commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+    profile = mlst.get("data", {}).get("profile", {})
+    # Check for multiple keys in the JSON file and define error message
+    keys = sorted(profile.keys())
+    original_key = keys[0] if keys else None
```
Could you clarify the purpose of this check, and if an error should be raised if the `key[0]` is not present? 

If keys['data'] (which is {}) if keys is a non-empty list of dict_keys else None?

sgsutcliffe replied:
There is an error below if no key is present in the MLST file. There needs to be at least one profile.

mattheww95 replied:
I see, and is there a specific reason to sort the keys? It may be unnecessary overhead. Also I beleive you can rely on the truthiness of an empty dictionary being False instead of this check.

But If I am missing something important please let me know and apologies!

apetkau replied:
@mattheww95 dictionaries don't have any particular order to keys, so without sorting it's possible that different runs could produce different results depending on order of keys.

mattheww95 replied:
I understand that part, I am just not quite understanding what this code block is actually testing. If the purpose is to just test that there are keys present in the dictionary e.g. that it is non-empty could the test not just something like `if profile: True`

apetkau replied:
The purpose is to test for the condition of multiple sample data within the JSON file. That is data for "SampleA" and "SampleB". If this is the case, it will sort the keys (sample names) and assume the first key (first sample name) is the appropriate data.

mattheww95 replied:
I see, Steven clarified the purpose for me (he described what you did) it makes sense now.

mattheww95 commented on tests/test_merge.py:
Could you add a test verifying an error report is created?

mattheww95 reviewed (changes requested):
Thanks so much for all the work @sgsutcliffe! I think I added too many comments, but they are just centered around removing magic numbers and assigning strings to variables.

apetkau reviewed (changes requested):
This looks great Steven. Thanks so much for your amazing work 😄. I've added my initial comments below.

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -198,18 +258,33 @@ def run_merge(config):
+            if error_message[2]:
+                output_error_file = outdir + "/" + error_message[0] + "_error_report.csv"
+                with open(output_error_file, "w", newline="") as f:
+                    writer = csv.writer(f)
```
Thanks for the context! I will change this

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -198,18 +258,33 @@ def run_merge(config):
+    #Write error messages for profile mismatch (compare_profiles())
+    for error_message in compare_error:
+            if error_message[2]:
+                output_error_file = outdir + "/" + error_message[0] + "_error_report.csv"
```
Thanks for the information -- with that context I agree and will make the changes.

sgsutcliffe replied:
[c9b0d94](https://github.com/phac-nml/locidex/pull/39/commits/c9b0d94c2d8a806af1e8ad1ca23fac704c6eae9f)

sgsutcliffe commented:
I remembered that for input_assure to be fully replicated in this added feature we also need to add the new MLST json files as output. 1d7ae8f

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+            sampledict[mlst_file] = [sample]
+        return sampledict
+
+def compare_profiles(mlst, sample_id, file_name):
```
I agree and I like validate_profile, I'll use that.

sgsutcliffe replied:
I decided to also rename a few other variables related to input_assure so it is more readible [9eb964c](https://github.com/phac-nml/locidex/pull/39/commits/9eb964ccc6b78768a2783d190472a3cd95641a01)

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -98,30 +100,37 @@ def check_files_exist(file_list: list[os.PathLike]) -> None:
 
     for f in file_list:
+        if key_sample_name:
+            alt_profile = key_sample_name[os.path.basename(f)][0]
```
Good point; still refamilairizing myself with python syntax. I'd thought something I was working on broke because the `f` in `file_list` was not available in that section of the function. Fixed!

sgsutcliffe replied:
Yes, `alt_profile` is the profile of the user has supplied to replace the profile currently in the MLST file, my other idea was `new_profile` but probably as ambiguous? I feel like something like `user_provided_profile` is too wordy, but I will go with that for now.

sgsutcliffe replied:
I made this change but it is now in the function `validate_input_file` [9eb964c](https://github.com/phac-nml/locidex/pull/39/commits/9eb964ccc6b78768a2783d190472a3cd95641a01)

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -98,30 +100,37 @@ def check_files_exist(file_list: list[os.PathLike]) -> None:
-                                                            db_version=db_version, 
-                                                            db_name=db_name, 
-                                                            perform_validation=perform_validation)     
+            if key_sample_name:
```
I took this and went a bit farther, I decided to take some of the code from `run_merge` and put it in there too. I also decided to rename a bunch of variables so that is more in line of profile validation.

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -167,14 +176,65 @@ def write_gene_fastas(seq_data,work_dir):
+    return files
+
+def read_samplesheet(sample_key_file):
+        f =open(sample_key_file,'r')
```
I did this and improved how it handles input, so that it can handle both types of samplesheets for gasnomenclature but I moved it into `read_file_list` function so that it doesn't repeat with each mlst file. 468f76e

sgsutcliffe commented:
Thanks @mattheww95 and @apetkau for the extremely helpful code review. It looked like simply using pandas cleared up a lot of the magic numbers, and I moved things around and renamed variables to hopefully improve the readibility of the code.

mattheww95 reviewed (changes requested):
Much better thank you for removing the magic @sgsutcliffe 😄 

I added some small requests to fix up, but I think I will have no further complaints after those changes are made.

apetkau commented on locidex/merge.py:
```diff
@@ -167,14 +187,73 @@ def write_gene_fastas(seq_data,work_dir):
+
+    return sampledict
+
+def validate_profiles(mlst, sample_id, file_name):
```
I am wondering if this could be changed to something like `validate_and_fix_profiles`?

My reasoning is that the passed `mlst` object will get modified by this function. Normally something named `validate` I would not expect to modify the inputs. But something with `fix` in the name I would. So the name change just makes this clearer that the input `mlst` gets modified.

apetkau commented on tests/test_merge.py:
```diff
@@ -1,39 +1,81 @@
+    assert same_mlst["data"]["profile"] == {'sample1': {'l1': '1', 'l2': '1', 'l3': '1'}}
+    assert mlst_report[2] == "sample1 ID and JSON key in sample1.mlst.json DO NOT MATCH. The 'sampleA' key in sample1.mlst.json has been forcefully changed to 'sample1': User should manually check input files to ensure correctness."
+
+def test_profile_validation_report(tmpdir):
```
Could you add a similar test case as this, but excluding the `profile_ref`? And then verify the sample names aren't changed/no `MLST_error_report.csv` was generated?

apetkau reviewed (changes requested):
This looks great @sgsutcliffe. Thanks so much for addressing my comments 😄 .

I have a few additional in-line comments for you below.

sgsutcliffe commented on tests/test_merge.py:
As we discussed elsewhere, `TemporaryDirectory` is an unused module, I decided to use pytests built-in `tmpdir` so I removed the vestige code from testing. 477790b

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -65,28 +67,33 @@ def get_file_list(input_files):
     return file_list
 
-def validate_input_file(data_in: dict, db_version: str, db_name: str, perform_validation: bool) -> tuple[ReportData, str, str]:
+def validate_input_file(data_in: dict, filename: str, db_version: str, db_name: str, perform_db_validation: bool, perform_profile_validation: bool, profile_refs_dict=dict) -> tuple[ReportData, str, str, list]:
```
Following up on the solution we discussed elsewhere: a79103b

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -4,6 +4,7 @@
 import re
 import sys
 import errno
+import csv
```
a79103b

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -167,14 +187,73 @@ def write_gene_fastas(seq_data,work_dir):
+
+
+    except:
+            logging.critical("File should be CSV that will contain either ['sample', 'mlst_alleles'] or ['sample_name', 'mlst_alleles']")
```
d66af08

sgsutcliffe commented on locidex/merge.py:
```diff
@@ -167,14 +187,73 @@ def write_gene_fastas(seq_data,work_dir):
+        if not match_status:
+            MLST_message = f"No key in the MLST JSON file ({file_name}) matches the specified sample ID '{sample_id}'. The first key '{original_key}' has been forcefully changed to '{sample_id}' and all other keys have been removed."
+            # Retain only the specified sample ID
+            mlst[data_key][profile_key] = {sample_id: profile.pop(original_key)}
```
12e2775

apetkau reviewed (approved):
Everything looks great to me. Thanks so much @sgsutcliffe 😄

mattheww95 reviewed (approved):
LGTM, Thanks so much for all the work!!!!! 🧙
